src/csharp/Grpc.Core/Interceptors/GenericInterceptor.cs

```diff
@@ -0,0 +1,449 @@
+#region Copyright notice and license
+
+// Copyright 2018 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core.Internal;
+
+namespace Grpc.Core.Interceptors
+{
+    /// <summary>
+    /// Provides a base class for generic interceptor implementations that raises
+    /// events and hooks to control the RPC lifecycle.
+    /// </summary>
+    public abstract class GenericInterceptor : Interceptor
```

Comment:
We could have fused the two interfaces together (i.e. `IInterceptableCallHandler` into `IServerCallHandler` and all of them are internal so it doesn't really make any difference. I think in particular I had the `Unimplemented` implementation in mind, but we can simply add a no-op method to that and merge the interfaces.

Reply:
Yes, but learning from Python, I feel like a standard `GenericInterceptor` being part of our library is valuable, as without it, most interceptors will be much more verbose to write. We can move it to the tests or examples now and maybe add it in a future implementation, but that will possibly fragment other interceptor libraries that will be written on top of it (which is probably fine). 

(for some reason Github shifted my comments between posts!)

Reply:
I have turned `GenericInterceptor` into an internal class for now, as it simplifies our own internal interceptor uses (for Metadata and in ClientBase), but it is good to leave it out for now as it will be a major API-stability commitment.